### PR TITLE
optimize encoder

### PR DESCRIPTION
**Pull request by josharian**
This is a grab-bag of optimizations.
I recommend reviewing commit-by-commit and rebasing instead of squashing.

Their cumulative effect, on my laptop, for an out-of-tree benchmark (sorry) is:

```
goos: darwin
goarch: arm64
pkg: bold.dev/tknz/gemma2b
cpu: Apple M3 Max
          │      a      │                  f                  │
          │   sec/op    │   sec/op     vs base                │
PureGo-16   3.494m ± 6%   2.642m ± 1%  -24.39% (p=0.000 n=15)

          │       a       │                  f                   │
          │     B/op      │     B/op      vs base                │
PureGo-16   1858.5Ki ± 0%   824.0Ki ± 0%  -55.66% (p=0.000 n=15)

          │       a       │                 f                  │
          │   allocs/op   │ allocs/op   vs base                │
PureGo-16   7574.000 ± 0%   4.000 ± 0%  -99.95% (p=0.000 n=15)
```

I also will understand if these are viewed as too intrusive/complicated for this codebase. :) I am happy to maintain a fork as needed.

### CODE REVIEW

**Comment by eliben** [comment]
Thanks, this is interesting. I'll find time to review these

I do wonder, however, about your use case. Can you elaborate where higher encoding performance is important to you, or is this just for fun? LLMs generally are pretty slow and in recent comparisons I've seen providers brag that their models reach something like 200 tokens/second (and that's probably on the beefiest HW). In comparison, unless I've messed up my benchmark, `go-sentencepiece` encodes at well over half a million tokens/sec, so it's hard for me to imagine a situation where this tokenization is a bottleneck.

**Comment by josharian** [comment]
> I'll find time to review these

Great, thanks. I made some decisions along the way with not 100% confidence. Pushback/questions are--always--welcome.

> Can you elaborate where higher encoding performance is important to you

As part of exploration and dataset preparation, I end up doing mass tokenization runs, so I really feel the performance--it directly impacts my iteration speed as I'm hacking around.

And I'm not working with frontier models (gemma, not gemini). And some of my work involves squeezing extra performance out, plus our use case is quite latency-sensitive.

I also generally have a habit of occasionally taking a day to crush any boxes I can in pprof. Cumulatively, that adds up.

Anyway, I put in this time to get to rough parity with the cgo implementation, at least for my use cases, so that I can switch to pure Go without guilt. Plus it was fun. :)

File: processor.go [hunk 1]
```diff
@@ -215,6 +215,23 @@ func (proc *Processor) Encode(text string) []Token {
 		return -1
 	})
 
+	buf := make([]byte, 64) // reusable buffer for merging symbols
+	mergeSymbols := func(x, y symListElem) (string, int, bool) {
+		sz := len(x.symbol) + len(y.symbol)
+		if cap(buf) < sz {
+			buf = append(buf[:0], make([]byte, sz)...)
```
Comment (eliben) [inline processor.go:222]: This seems a bit obscure without a comment.
I don't think tokens larger than 64 bytes are very likely anyway (and the 64 number can probably be increased without harm?), so would it really hurt performance to have a more straightforward reallocation here? (assign larger buf)

File: processor.go [hunk 2]
```diff
@@ -215,6 +215,23 @@ func (proc *Processor) Encode(text string) []Token {
...
+		}
+		buf = buf[:cap(buf)]
```
Comment (eliben) [inline processor.go:224]: Why is this line needed?

**Review by eliben — COMMENTED** [review]
Some questions about the string allocation cache commit

File: internal/priorityqueue/priorityqueue.go [hunk 3]
```diff
@@ -44,10 +44,42 @@ func (pq *PriorityQueue[T]) PopMax() T {
...
+	for ; i < len(pq.items); i++ {
+		if rm(pq.items[i]) {
+			break
+		}
+	}
+	if i == len(pq.items) {
+		return // nothing to remove
+	}
+	for j := i + 1; j < len(pq.items); j++ {
+		v := pq.items[j]
```
Comment (eliben) [inline internal/priorityqueue/priorityqueue.go:65]: Nit: please rewrite this w/o continue, e.g. body in `if v := pq.items[j]; !rm(v) {`...

File: internal/priorityqueue/priorityqueue.go [hunk 4]
```diff
@@ -44,10 +44,42 @@ func (pq *PriorityQueue[T]) PopMax() T {
...
 	pq.items = pq.items[:len(pq.items)-1]
-	pq.siftdown()
+	pq.siftdown(1)
 	return maxItem
 }
 
+// RemoveFunc removes all elements for which rm returns true.
+func (pq *PriorityQueue[T]) RemoveFunc(rm func(T) bool) {
+	// This is effectively slices.DeleteFunc, but inlined because we start from index 1,
+	// and we don't need to clear the tail.
```
Comment (eliben) [inline internal/priorityqueue/priorityqueue.go:54]: Why don't we need to clear the tail?

File: internal/priorityqueue/priorityqueue.go [hunk 5]
```diff
@@ -19,8 +19,8 @@ type PriorityQueue[T any] struct {
 // compares the priorities of two items a and b; it should return a number > 0
 // if the priority of a is higher, 0 if the priorities are equal a number < 0
 // otherwise.
-func New[T any](cmp func(a, b T) int) *PriorityQueue[T] {
-	return &PriorityQueue[T]{cmp: cmp, items: make([]T, 1)}
+func New[T any](sz int, cmp func(a, b T) int) *PriorityQueue[T] {
```
Comment (eliben) [inline internal/priorityqueue/priorityqueue.go:22]: Call it `sizeHint`, if -1 is passed it should be ignored, and the comment on `New` should explain this

File: processor.go [hunk 6]
```diff
@@ -140,40 +140,38 @@ func NewProcessor(protoReader io.Reader) (*Processor, error) {
 	}, nil
 }
 
+type symListElem struct {
+	symbol string
+	// noMerge means this is a user-defined symbol we're not allowed to merge with neighbors.
+	noMerge bool
+}
+
+func (e symListElem) Dead() bool {
```
Comment (eliben) [inline processor.go:149]: I don't love this commit... since it doesn't affect run-time, let's back it out for now. I'd like to think a bit more about memory management in Encode in general

File: processor.go (hunk 1 above)
Comment (josharian) [reply processor.go:222]: Will do. Old habits die hard. :)

One could precalculate the longest string piece when loading and store that length in the processor, and simplify all this even more.

What do you think?

File: processor.go (hunk 2 above)
Comment (josharian) [reply processor.go:224]: For the map lookup, we do:

```go
buf = buf[:sz]
```

but copy respects slice length, so this might make buf too small. So we reset buf to be as long as its cap before copying.

File: internal/priorityqueue/priorityqueue.go (hunk 3 above)
Comment (josharian) [reply internal/priorityqueue/priorityqueue.go:65]: as you wish

File: internal/priorityqueue/priorityqueue.go (hunk 4 above)
Comment (josharian) [reply internal/priorityqueue/priorityqueue.go:54]: There are no pointers in it, so it doesn't impact GC. (slices.DeleteFunc doesn't know one way or the other.) And nobody will ever read the tail. So clearing the tail is just wasted work.

File: internal/priorityqueue/priorityqueue.go (hunk 5 above)
Comment (josharian) [reply internal/priorityqueue/priorityqueue.go:22]: OK

File: processor.go (hunk 6 above)
Comment (josharian) [reply processor.go:149]: OK

**Comment by josharian** [comment]
Will push up a revised copy once the questions above are answered. Sorry for the delay, been a busy couple of days.

File: processor.go (hunk 1 above)
Comment (eliben) [reply processor.go:222]: Yes, that sounds good!

File: internal/priorityqueue/priorityqueue.go (hunk 4 above)
Comment (eliben) [reply internal/priorityqueue/priorityqueue.go:54]: I must be missing something, but why do you say it has no pointers? T can be a pointer

File: internal/priorityqueue/priorityqueue.go (hunk 5 above)
Comment (josharian) [reply internal/priorityqueue/priorityqueue.go:22]: Done

File: internal/priorityqueue/priorityqueue.go (hunk 4 above)
Comment (josharian) [reply internal/priorityqueue/priorityqueue.go:54]: Oops. Done.

File: internal/priorityqueue/priorityqueue.go (hunk 3 above)
Comment (josharian) [reply internal/priorityqueue/priorityqueue.go:65]: Done

File: processor.go (hunk 6 above)
Comment (josharian) [reply processor.go:149]: Done

File: processor.go (hunk 1 above)
Comment (josharian) [reply processor.go:222]: Done

**Review by josharian — COMMENTED** [review]
PTAL
